worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala

```diff
@@ -526,4 +530,278 @@ class PushDataHandler extends BaseMessageHandler with Logging {
+    val (workerSourceMaster, workerSourceSlave) =
+      messageType match {
+        case Type.PUSH_DATA_HAND_SHAKE =>
+          (WorkerSource.MasterPushDataHandshakeTime, WorkerSource.SlavePushDataHandshakeTime)
+        case Type.REGION_START =>
+          (WorkerSource.MasterRegionStartTime, WorkerSource.SlaveRegionStartTime)
+        case Type.REGION_FINISH =>
+          (WorkerSource.MasterRegionFinishTime, WorkerSource.SlaveRegionFinishTime)
+      }
+
+    val location = isMaster match {
+      case true => partitionLocationInfo.getMasterLocation(shuffleKey, partitionUniqueId)
+      case false => partitionLocationInfo.getSlaveLocation(shuffleKey, partitionUniqueId)
+    }
+    workerSource.startTimer(if (isMaster) workerSourceMaster else workerSourceSlave, s"$requestId")
+    val wrappedCallback =
+      new WrappedRpcResponseCallback(
+        messageType,
+        isMaster,
+        requestId,
+        null,
+        location,
+        if (isMaster) workerSourceMaster else workerSourceSlave,
+        callback)
+
+    val isReturn = checkLocationNull(
+      messageType,
+      shuffleKey,
+      partitionUniqueId,
+      null,
+      location,
+      callback,
+      wrappedCallback)
+    if (isReturn) return
+
+    val (isReturnWriter, fileWriter) =
+      getFileWriterAndCheck(messageType, location, isMaster, callback)
+    if (isReturnWriter) return
+
+    if (isCheckSplit) {
+      val isReturnDisk = checkDiskFullAndSplit(fileWriter, isMaster, null, callback)
+      if (isReturnDisk) return
+    }
+
+    try {
+      messageType match {
+        case Type.PUSH_DATA_HAND_SHAKE => {
+          fileWriter.pushDataHandShake(message.asInstanceOf[PushDataHandShake].numPartitions)
+        }
+        case Type.REGION_START => {
+          fileWriter.regionStart(
+            message.asInstanceOf[RegionStart].currentRegionIndex,
+            message.asInstanceOf[RegionStart].isBroadcast)
+        }
+        case Type.REGION_FINISH => {
+          fileWriter.regionFinish()
+        }
+      }
+      // for master, send data to slave
+      if (location.getPeer != null && isMaster) {
+        // to do replica
+        wrappedCallback.onSuccess(ByteBuffer.wrap(Array[Byte]()))
+      } else {
+        wrappedCallback.onSuccess(ByteBuffer.wrap(Array[Byte]()))
+      }
+    } catch {
+      case t: Throwable =>
+        callback.onFailure(new Exception(s"$messageType failed", t))
+    }
+  }
+
+  class WrappedRpcResponseCallback(
+      messageType: Message.Type,
+      isMaster: Boolean,
+      requestId: Long,
+      softSplit: AtomicBoolean,
+      location: PartitionLocation,
+      workerSourceTime: String,
+      callback: RpcResponseCallback)
+    extends RpcResponseCallback {
+    override def onSuccess(response: ByteBuffer): Unit = {
+      if (isMaster) {
+        workerSource.stopTimer(workerSourceTime, s"$requestId")
```

Comment:
The stop timer method always needs to be called. It can moved out of if block.

Reply:
it has been removed outside